libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1263,7 +1263,39 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 				// as require with custom errors is not supported in legacy codegen.
 				auto const* magicType = dynamic_cast<MagicType const*>(arguments[1]->annotation().type);
 				if (magicType && magicType->kind() == MagicType::Kind::Error)
-					solUnimplemented("Require with a custom error is only available using the via-ir pipeline.");
+				{
+					// Make sure that error constructor arguments are evaluated regardless of the require condition
+					auto const& errorConstructorCall = dynamic_cast<FunctionCall const&>(*arguments[1]);
+					errorConstructorCall.expression().accept(*this);
+					std::vector<Type const*> argumentTypes{};
+					unsigned stackDepth{0};
+					for (ASTPointer<Expression const> const& arg: errorConstructorCall.sortedArguments())
+					{
+						arg->accept(*this);
+						stackDepth += arg->annotation().type->sizeOnStack();
+						argumentTypes.push_back(arg->annotation().type);
+					}
+					// stack: <condition> <arg0> <arg1> ... <argN>
+					// Move condition to the top of the stack
+					utils().moveIntoStack(arguments.at(0)->annotation().type->sizeOnStack(), stackDepth);
+					// stack: <arg0> <arg1> ... <argN> <condition>
+					m_context << Instruction::ISZERO << Instruction::ISZERO;
+					auto success = m_context.appendConditionalJump();
+
+					auto const* errorDefinition = dynamic_cast<ErrorDefinition const*>(ASTNode::referencedDeclaration(errorConstructorCall.expression()));
+					solAssert(errorDefinition && errorDefinition->functionType(true));
+
+					utils().revertWithError(
+						errorDefinition->functionType(true)->externalSignature(),
+						errorDefinition->functionType(true)->parameterTypes(),
+						argumentTypes
+					);
+					m_context.adjustStackOffset(static_cast<int>(stackDepth));
+					m_context << success;
+					for (ASTPointer<Expression const> const& arg: errorConstructorCall.sortedArguments())
+						utils().popStackElement(*arg->annotation().type);
+					break;
```

Comment:
By the way, for a moment, I had half a mind to suggest to still replace all of this with a simple
```C++
		// stack: <condition> <arg0> <arg1> ... <argN>
		m_context.callYulFunction(m_context.utilFunctions().requireWithErrorFunction(errorConstructorCall), stackDepth, 0);
```
after all :-). However: I'd need to think through this, but we would probably need to annoyingly invert the order of arguments for the call (a Yul function expects arguments on stack right-most first left-most last...)... which is quite bad - the function then would consume them again in the opposite order and we probably won't get inlining and stuff, so maybe actually better to keep it as is...
Are you sure that's such a bad idea ;-)?

Reply:
Having opposite evaluation orders between Solidity and Yul/EVM is really crazy...

Reply:
Then again, our calling convention means we also have a huge ``moveIntoStack`` in ``CompilerContext::callYulFunction`` - taken together they'd probably cancel out and the libevmasm optimizer would even do that :-). But yeah, if you're curious you can try and check gas costs, but it's likely better to keep it as is in any case.

Reply:
At this point I'm leaving it as is, especially since I've already taken a look between the generated asm by the legacy vs via-ir pipeline, and as you said, the IR pipeline also performs  the same swaps (moveIntoStack) as the legacy, so gas costs would likely remain the same (or close enough).

The only difference is the lack of the small reverts optimization that we have in the IR pipeline, but that's just another reason to switch to it.